src/lib/stable/canister_methods/index.ts

```diff
@@ -62,104 +80,151 @@ export function decoratorArgumentsHandler<This, Args extends any[], Return>(
+                ['query']
+            );
+        }
 
-    if (canisterMethodMode === 'inspectMessage') {
-        globalThis._azleMethodMeta.inspect_message = {
-            name,
-            index
-        };
-    }
+        if (canisterMethodMode === 'update') {
+            this._azleMethodMeta.updates?.push({
+                name,
+                index
+            });
 
-    globalThis._azleCallbacks[indexString] = async (
-        args?: Uint8Array
-    ): Promise<void> => {
-        try {
-            await executeAndReplyWithCandidSerde(
-                canisterMethodMode,
-                args ?? new Uint8Array(),
-                originalMethod.bind(globalThis._azleCanisterClassInstance),
+            this._azleCanisterMethodIdlTypes[name] = IDL.Func(
                 paramIdlTypes ?? [],
-                returnIdlType,
-                options?.manual ?? false
+                returnIdlType === undefined ? [] : [returnIdlType]
             );
-        } catch (error: any) {
-            handleUncaughtError(error);
         }
-    };
+
+        if (canisterMethodMode === 'init') {
+            this._azleMethodMeta.init = {
+                name,
+                index
+            };
+
+            this._azleInitAndPostUpgradeIdlTypes.push(
+                IDL.Func(paramIdlTypes ?? [], [], ['init'])
+            );
+        }
+
+        if (canisterMethodMode === 'postUpgrade') {
+            this._azleMethodMeta.post_upgrade = {
+                name,
+                index
+            };
+
+            this._azleInitAndPostUpgradeIdlTypes.push(
+                IDL.Func(paramIdlTypes ?? [], [], ['post_upgrade'])
+            );
+        }
+
+        if (canisterMethodMode === 'preUpgrade') {
+            this._azleMethodMeta.pre_upgrade = {
+                name,
+                index
+            };
+        }
+
+        if (canisterMethodMode === 'heartbeat') {
+            this._azleMethodMeta.heartbeat = {
+                name,
+                index
+            };
+        }
+
+        if (canisterMethodMode === 'inspectMessage') {
+            this._azleMethodMeta.inspect_message = {
+                name,
+                index
+            };
+        }
+
+        this._azleCallbacks[indexString] = async (
+            args?: Uint8Array
+        ): Promise<void> => {
+            try {
+                await executeAndReplyWithCandidSerde(
+                    canisterMethodMode,
+                    args ?? new Uint8Array(),
+                    originalMethod.bind(this), // TODO manually test context and ensure we have automatic tests for this
+                    paramIdlTypes ?? [],
+                    returnIdlType,
+                    options?.manual ?? false
+                );
+            } catch (error: any) {
+                handleUncaughtError(error);
+            }
+        };
+
+        globalThis._azleExportedCanisterClassInstance = this;
```

Comment:
If you have multiple classes won't this result in the last class executed being the class that represents the canister?

Reply:
Or will the default exported class always be the last to finish up since this happens at the end and any other class that might be initialized in pursuit of initializing this class will all be called and done before we get to this line and the default exported class will always ultimately be the last to write to this variable?

Reply:
Addressed, please scrutinize deeply my solution, considering global state issues as multiple instances might be created during class constructor or variable instantiation